src/EditorFeatures/CSharp/ConvertNamespace/ConvertNamespaceCommandHandler.cs

```diff
@@ -0,0 +1,168 @@
+
+            // Otherwise, make a transaction for the edit and replace the buffer with the final text.
+            using var transaction = CaretPreservingEditTransaction.TryCreate(
+                this.DisplayName, args.TextView, _textUndoHistoryRegistry, _editorOperationsFactoryService);
+
+            var edit = args.SubjectBuffer.CreateEdit(EditOptions.DefaultMinimalChange, reiteratedVersionNumber: null, editTag: null);
+            edit.Replace(new Span(0, args.SubjectBuffer.CurrentSnapshot.Length), convertedRoot.ToFullString());
+
+            edit.Apply();
+
+            // Attempt to place the caret right after the semicolon of the file-scoped namespace.
+            var annotatedToken = convertedRoot.GetAnnotatedTokens(s_annotation).FirstOrDefault();
+            if (annotatedToken != default)
+                args.TextView.Caret.MoveTo(new SnapshotPoint(args.SubjectBuffer.CurrentSnapshot, annotatedToken.Span.End));
+
+            transaction?.Complete();
+        }
+
+        /// <summary>
+        /// Returns true if semicolon is typed after a namespace name that should be converted.
+        /// </summary>
+        private static CompilationUnitSyntax? ConvertNamespace(
+            TypeCharCommandArgs args,
+            CommandExecutionContext executionContext)
+        {
+            if (args.TypedChar != ';' || !args.TextView.Selection.IsEmpty)
+                return null;
+
+            var subjectBuffer = args.SubjectBuffer;
+            var caretOpt = args.TextView.GetCaretPoint(subjectBuffer);
+            if (!caretOpt.HasValue)
+                return null;
+
+            var caret = caretOpt.Value.Position;
+            var document = subjectBuffer.CurrentSnapshot.GetOpenDocumentInCurrentContextWithChanges();
+            if (document == null)
+                return null;
+
+            var cancellationToken = executionContext.OperationContext.UserCancellationToken;
+            var root = (CompilationUnitSyntax)document.GetRequiredSyntaxRootSynchronously(cancellationToken);
+
+            // User has to be *after* an identifier token.
+            var token = root.FindToken(caret);
+            if (token.Kind() != SyntaxKind.IdentifierToken)
+                return null;
+
+            if (caret < token.Span.End ||
+                caret >= token.FullSpan.End)
+            {
+                return null;
+            }
+
+            var namespaceDecl = token.GetRequiredParent().GetAncestor<NamespaceDeclarationSyntax>();
+            if (namespaceDecl == null)
+                return null;
+
+            // That identifier token has to be the last part of a namespace name.
+            if (namespaceDecl.Name.GetLastToken() != token)
+                return null;
+
+            // Pass in our special options, and C#10 so that if we can convert this to file-scoped, we will.
+            if (!ConvertNamespaceAnalysis.CanOfferUseFileScoped(s_optionSet, root, namespaceDecl, forAnalyzer: true, LanguageVersion.CSharp10))
```

Comment:
I'm on my phone, so might have missed it, but why pass in C# 10? Shouldn't we pass in the projects version so we don't try to convert for older langversions?

Reply:
so, the idea here is that if a user types `;` in this location, there is only one thing they're trying to do (use the new file-scoped-namespace feature).  We shouldn't punish them by making it not work just because htey haven't updated their lang version.  We should just make the change, and the user can then move to C#10 at that point.

Reply:
Same question here

Reply:
> We should just make the change, and the user can then move to C#10 at that point.

This seems like a new direction I haven't seen us take before. Why does this not apply to all of the other fixers and refactorings we offer? If a user accidentally types a semicolon in a namespace on their .NET Framework 4.5 project, moving to C# 10 could be very non-trivial.

Also, why doesn't this logic apply to the two namespaces, or nested namespaces scenario? If the reasoning is "the user typed a semicolon, they must want file scoped namespaces", then why not also just convert the first namespace in the file, and the user can move the second one to another file afterwards?

Reply:
> This seems like a new direction I haven't seen us take before.

@davidwengier Not completely a new direction (see  https://github.com/dotnet/roslyn/issues/55369), but it might be debatable.

Reply:
> If a user accidentally types a semicolon in a namespace on their .NET Framework 4.5 project, moving to C# 10 could be very non-trivial.

I think tehre's a misunderstanding here.  This is not about an errant semicolon typed at some particular place.  This is about going up to an existing namespace (which has braces, and members) going right after the name, and typing 'semicolon' :)

Reply:
> Why does this not apply to all of the other fixers and refactorings we offer?

It's my opinion here that this is a very rare, but explicit, trigger that indicates the user intends to use a file-scoped-namespace.  The trigger specificity, and the lack of overlap with any other constructs means we can specialize this as something explicit we do.

>Also, why doesn't this logic apply to the two namespaces, or nested namespaces scenario? If the reasoning is "the user typed a semicolon, they must want file scoped namespaces", then why not also just convert the first namespace in the file, and the user can move the second one to another file afterwards?

This is a good question.  Primarily I was concerned about actually breaking code and putting the user in a space that they didn't know how to fix.  Namely that they then might trigger "move to file" and things might go wrong because it would take the new *nested* interpretation instead of hte original sibling interpretation.

We also checked the C# repos we have access to and found that "single namespace + type" is the 99.9 case (i think 99.97), so it felt this was hitting a reasonable sweet spot here :)

Reply:
I'm not misunderstanding, I just know there is a possibility that I would click on a namespace, want to add something to the end of it to make it more specific, and automatically hit semi-colon when I get the end.

Reply:
Ok. That's a reasonable point. I think it feels unlikely, but I'm amenable to gating this. Thanks!

Reply:
ok.  we're gated behind the option now.